rpc/rest/security/rest_security.go

```diff
@@ -0,0 +1,422 @@
+	router    *mux.Router
+	formatter *render.Render
+
+	// User database keeps all known users with permissions and hashed password. Users are loaded from
+	// HTTP config file
+	// TODO add option to register users
+	userDb AuthenticationDB
+	// Permission database is a map of name/permissions and bound URLs
+	groupDb map[string][]*access.PermissionGroup_Permissions
+
+	// Token claims
+	expTime time.Duration
+}
+
+// NewAuthenticator prepares new instance of authenticator.
+func NewAuthenticator(router *mux.Router, ctx *Settings, log logging.Logger) AuthenticatorAPI {
+	a := &authenticator{
+		router: router,
+		log:    log,
+		formatter: render.New(render.Options{
+			IndentJSON: true,
+		}),
+		groupDb: make(map[string][]*access.PermissionGroup_Permissions),
+		expTime: ctx.ExpTime,
+	}
+
+	// Authentication store
+	if ctx.AuthStore != nil {
+		a.userDb = ctx.AuthStore
+	} else {
+		a.userDb = CreateDefaultAuthDB()
+	}
+
+	// Set token signature
+	signature = ctx.Signature
+	if a.expTime == 0 {
+		a.expTime = defaultExpTime
+		a.log.Debugf("Token expiration time claim not set, defaulting to 1 hour")
+	}
+
+	// Process users in go routine, since hashing may take some time
+	go func() {
+		// Add admin-user, enabled by default, always has access to every URL
+		hash, err := bcrypt.GenerateFromPassword([]byte("ligato123"), ctx.Cost)
+		if err != nil {
+			a.log.Errorf("failed to hash password for admin: %v", err)
+		}
+		a.userDb.AddUser(admin, string(hash), []string{admin})
+
+		for _, user := range ctx.Users {
+			if user.Name == admin {
+				a.log.Errorf("rejected to create user-defined account named 'admin'")
+				continue
+			}
+			if err := a.userDb.AddUser(user.Name, user.PasswordHash, user.Permissions); err != nil {
+				a.log.Errorf("failed to add user %s: %v", user.Name, err)
+				continue
+			}
+			a.log.Debug("Registered user %s, permissions: %v", user.Name, user.Permissions)
+		}
+	}()
+
+	// Admin-group, available by default and always enabled for all URLs
+	a.groupDb[admin] = []*access.PermissionGroup_Permissions{}
+
+	a.registerSecurityHandlers()
+
+	return a
+}
+
+// AddPermissionGroup adds new permission group.
+func (a *authenticator) AddPermissionGroup(group ...*access.PermissionGroup) {
+	for _, newPermissionGroup := range group {
+		if _, ok := a.groupDb[newPermissionGroup.Name]; ok {
+			a.log.Warnf("permission group %s already exists, skipped")
+			continue
+		}
+		a.log.Debugf("added HTTP permission group %s", newPermissionGroup.Name)
+		a.groupDb[newPermissionGroup.Name] = newPermissionGroup.Permissions
+	}
+}
+
+// Validate the request
+func (a *authenticator) Validate(provider http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// Token may be accessed via cookie, or from authentication header
+		var tokenString string
+		// Try to read the cookie first
+		cookie, err := req.Cookie(cookieName)
```

Comment:
This should be as fallback not as first choice.

Reply:
And perhaps just extract this retrieval of token from request to separate function.

Reply:
I put the cookie check first because it can return only "not found" error, but since we don't know whether the user is accessing URL via browser or curl, we can't return the error. 
Also, I think it does not matter what we try first.

But I put the code to the separate method.

Reply:
> we don't know whether the user is accessing URL via browser or curl

This is simply not true. Every HTTP client will sends information about "user agent" used inside the HTTP header `User-Agent`. For curl it will print `curl/X.XX.X`. You can retrieve it using `req.Header.Get("User-Agent")`. 

However such check should not be needed here! The most common way to use JWT is with Authorization header and it should always come first. What if cookies are stale and you are sending correct info through Authorization header? Then until you delete cookies it won't work. 

Authorization header should always be checked first as the main source of thruth. Cookies are just some crutch, they are not meant to be used for security related things at all, but they can serve good enough fallback here so we cant easily use this in the browsers without coding Javascript clients.